### ISSUE TITLE
Centralized seed services, inspired by WebSeeds

### ISSUE THREAD

**Issue by mikeal** [ipld/ipld#57]
First, some background.

### Bittorrent WebSeeds

Bittorrent has this great feature called ["WebSeeds."](http://www.bittorrent.org/beps/bep_0019.html) It's rather simple, instead of having to spin up a service of bittorrent nodes in order to keep content alive you can simply add an HTTP or FTP URL as the fallback location of the content.

Bittorrent clients try to pull content from other peers but if none are available, or nobody has a complete copy of the file, or if they are just too slow, the client has the option of pulling the content out of the centralized service.

This allows people to keep content up much more simply that they would otherwise be able to if they had to manage a cluster of bittorrent nodes, without sacrificing the other benefits of bittorrent when the content is popular and the network around it  is healthy.

### Centralized Block and Graph services

The data-structures in IPLD offer some big benefits and upgrades compared to bittorrent. Instead of associating a specific URL with a specific file we could actually have fallback **services** which are known to have large caches of IPLD blocks.

These services could provide not just a fallback but also get us out around certain performance penalties in IPFS/IPLD. For instance, take the simple case of pulling up a website for the first time using IPNS/IPFS:

* IPNS resolves to a CID.
* IPFS looks in the DHT for that CID and establishes a network.
* As IPFS connects to peers it begins to pull through the graph to grab the content.

This is *always* slower for the first load than centralized solutions because it takes longer to establish the network to retrieve that content than it would to just connect to a central server.

However, if you could configure a block/graph service in IPFS it could look more like this:

* IPNS resolves to a CID.
* In parallel:
  * IPFS looks in the DHT for that CID and establishes a network.
  * Existing HTTP2 connections to block/graph services are queried for the CID.

Now you're already pulling in content while the network is established and you can continue to parallize/optimize grabbing the content as the connections are made. This is the best of both worlds and could actually beat the performance of existing web fetches.

Applications that use IPFS could either configure a set of known services in the node or provide a list of seed services when they ask for specific pieces of content.

I'm thinking that there are two distinct sets of services:

* Block Service (stores content by multihash)
* Graph Service (stores content by CID, returns the Block data **and** meta information that includes whether or not the service contains the full graph referenced by the CID)

If something like this was available it would make a lot of our infrastructure challenges a lot simpler. We could define a very simple REST API that is more or less compatible with S3 and people could literally just stick Cloudflare in front of it for global caching.

Yes, this is a centralized service, but hosted IPFS clusters are **also** centralized services, they are just available in the DHT. As long as the data-structures are content-addressed centralized services are no more than a caching layer and do about as much to "centralize" the data-structure as an offline cache does.

Thoughts? @daviddias @alanshaw @olizilla @vmx @eefahy

**Comment by vmx** [comment]
Without thinking too hard about all implications, it sounds good to me.

**Comment by eefahy** [comment]
If I understand the proposal correctly, once the request comes in we are essentially parallelizing the data fetch via IPFS and one of these backend services? That sounds pretty great to me. One concern would be that we deemphasize our need to make bitswap more performant and would complicate our metrics on bitswap performance on the gateway (which we don't actually have implemented yet)

**Comment by mikeal** [comment]
> One concern would be that we deemphasize our need to make bitswap more performant and would complicate our metrics on bitswap performance on the gateway

Agreed on the concerns about performance, but I think metrics we'd get from the block services could compensate for any lost gateway metrics. We could even estimate decent metrics in the p2p network we have a hard time measuring if we just look at the difference between the "top of graph" pulled from the block service and the rest of the graph which we can assume they are pulling from a p2p network.

**Comment by Stebalien** [comment]
> I'd also love to reduce the need to wait for DNS propagation when the CID is changed.

?

---

This:

> It's a far more accessible solution to average developers today than running a cluster. 

Is in conflict with:

> Instead of associating a specific URL with a specific file we could actually have fallback services which are known to have **large caches** of IPLD blocks.

The former is concerned with small developers who just want to store *their* data with Amazon. The latter sounds more like an "IPLD cache as a service" system (that could totally run an IPFS cluster [with a click](https://azuremarketplace.microsoft.com/en-us/marketplace/apps/microsoft-azure-blockchain.azure-blockchain-ipfs)).

The former would likely be configured on a per-dapp basis (a dapp would configure a "backup" block exchange that just fetches blocks from S3). The latter would be well-known a known centralized service which, IMO, doesn't really need to exist given bitswap.

Note: the former isn't going to help with things like page loading because we'd need to know *which* s3 bucket to use *before* we start loading the page.

---

Let's try to hone in on the problem a bit. What's the UX or DX flow you're trying to enable?

**Comment by Stebalien** [comment]
So, what if we instead we push features that better enable pinning services? A web-backed blockservice will work for individual apps but makes it *much* harder for apps to interoperate unless said blockservice *also* runs an IPFS node, making it's blocks available over bitswap.

**Comment by Stebalien** [comment]
Yeah, but if data gets soloed into per-app S3 buckets, we're not building dapps. We're going to have a *very* hard time convincing devs to use IPLD if we're not going to give them something substantially better than firebase. Worse, I'm worried the system will end up *depending* on these per-app shared blockstores.

**Comment by mikeal** [comment]
> Yeah, but if data gets soloed into per-app S3 buckets, we're not building dapps.

The data isn't *completely* silo'd, any data that is in use by connected peers will be in the DHT. Essentially, one "mega-peer" is not in the DHT and not available to Bitswap. This is already a better case than we have with forked DHT's, where none of the connected peers for that app are sharing data in the mainline DHT.

Sure, saying "you have to run a cluster to have reliable data" will make sure all the data used is available but it also means that many applications simply won't be written as a result, so the data from those applications is still never going to be accessible.

But I think there's a way we can still avoid these segmentation concerns with graph/block services. If we do not enable the second API I suggested and **only** enable the node level configuration `ipfs.config('blockservice.myapp', URL)` we could find ways to use these nodes as proxies for **all** the data in the services and not just what is locally available in each nodes.

**Comment by mikeal** [comment]
Additional note for readers: I detailed the gap between the current S3 storage backend and what this is proposing here https://github.com/protocol/pl-ipfs-team/issues/24#issuecomment-457343740

**Comment by eocarragain** [comment]
@mikeal that link seems to be to a private repo. Can you repeat the information here?

**Comment by mikeal** [comment]
@eocarragain apologies, here it is:

---

The problem with our *current* S3 backed storage is that it doesn't distinguish between read and write configuration. In the most common use cases you want:

* A single provider writing data that is stored long term (the application provider).
* Many people reading from long term (non-DHT discovered) storage.
* Many people keeping local copies of the data in their own storage.
* Bonus Points:
  * Content that is currently being stored by the peer network should be preferred over long-term storage.

**Comment by parkan** [comment]
also: S3 is not the sole backend type possible (Google Cloud Storage, DigitalOcean Storage, owndatacenter, etc)

using the S3 API has certain advantages, but the least common denominator for access is probably HTTP

**Comment by mikeal** [comment]
Yup, and all of these services have a similar enough HTTP interface that we could do the same thing across all of them if it’s read-only since only authentication differs between them.

**Comment by aschmahmann** [comment]
While I see the utility of having a centralized service you could fall back on for getting content, it feels strange to be running these two systems in parallel and requiring end-user software to have preconfigured centralized services to use for every application (e.g. `ipfs.config('blockservice.myapp', URL)`).

An alternative that's close to this, but perhaps a little more IPFS friendly, would be to allow using Multiaddrs to indicate content routing over non-libp2p protocols (libp2p/notes#11). In particular, the flow from CID to data is:

**Currently**
1. Search DHT for CID
2. Receive a list of multiaddrs that have advertised that they have the CID
3. Use a mutually supported exchange protocol (i.e. bitswap) to retrieve the data from the libp2p nodes at the advertised multiaddrs

**With HTTP Support**
1. Search the DHT for CID
2. Receive a list of multiaddrs that have advertised that they have the CID
3. Use the multiaddrs to determine which protocol should be used to retrieve the data (i.e. `ip4/1.2.3.4/tcp/4001/p2p/PeerID` for libp2p protocols like bitswap, `/dns4/mybucket.s3.amazonaws.com/tcp/80/http/DataFolder/DataBlock` for downloading data over HTTP)

Yes, someone still needs to run an IPFS node that submits Provider records pointing at the HTTP addresses, however, the node no longer needs 24/7 up time. Additionally, we could potentially allow anyone to publish these HTTP-based Provider records allowing us to effectively draft content into IPFS that is currently only accessible via the centralized web.

**Comment by rvagg** [comment]
@aschmahmann:

> the node no longer needs 24/7 up time. Additionally, we could potentially allow anyone to publish these HTTP-based Provider records allowing us to effectively draft content into IPFS that is currently only accessible via the centralized web

I don't think I know enough yet about discovery but is there an attack vector here? Is it a cheaper operation to poison the DHT with pointers to content that you don't host yourself and don't have to take responsibility for not existing than advertising that you have a certain CID?

**Comment by aschmahmann** [comment]
@rvagg I could be mistaken, but I don't think there is any new attack on the DHT node storing the provider record.

The DHT nodes never ask for any sort of "proof" that you have the data (e.g. probabilistically ask advertisers to send the data, ask for the data hashed with some nonce, etc.), so you could easily not have the data. This is on top of the issue that even if the provider node has the data there's no way to know they'll be able to/want to provide it at an acceptable upload speed. You're basically just trusting the advertising nodes off of their reputation (we could add various types of reputation systems going forward, but currently I think we trust all nodes equally).

However, one attack that we do open up if we allow *any* node to advertise on behalf of another party is a DoS attack on the other party. For example, if I know QmABC is really popular I could setup a DoS on http://InnocentBystander.com/index.html by just advertising `/dns4/InnocentBystander.com/tcp/80/http/index.html/` for QmABC. However, if QmABC has other hosts (not an unreasonable assumption for popular content) then the requests will be more distributed and hopefully not severely overwhelm InnocentBystander.com. We could (whether we allow arbitrary nodes to advertise on behalf of others or not) allow users to report unavailable/bad providers so they get evicted from the DHT - although deciding who to trust is essentially a reputation system problem.

It's also worth noting that if we are nonetheless still concerned about random advertisers then we could potentially require that the advertising nodes be "certified" by the non-libp2p service (e.g. signed DNS records with a field for the advertiser). However, I'm not currently convinced this is necessary.

**Comment by lidel** [comment]
Related:  [`?format=block` and `?format=car`](https://github.com/ipfs/go-ipfs/issues/8234) could effectively enable every public gateway to be used as block/DAG archive provider.
